src/lib/integration_test_local_engine/mina_docker.ml

```diff
@@ -1091,7 +1096,20 @@ module Network_manager = struct
     let%bind () = File_system.remove_dir t.docker_dir in
     Deferred.unit
 
-  let destroy t =
+  let tear_down t network =
+    if t.generate_code_coverage then
+      let open Malleable_error.Let_syntax in
+      let coverage_manager = Coverage_manager.create ~logger:t.logger in
+      let%bind _ =
+        Coverage_manager.download_coverage_data_from_containers coverage_manager
+          ~nodes:(Docker_network.all_nodes network)
+      in
+      Malleable_error.ok_unit
+    else Malleable_error.ok_unit
+
+  let destroy t network =
+    let open Malleable_error.Let_syntax in
+    let%bind () = tear_down t network in
```

Comment:
Using a `Malleable_error` here means that the deployment *won't* be destroyed if you can't dump coverage data. That would be bad.